packages/compass-shell/src/plugin.js

```diff
@@ -1,40 +1,51 @@
-import React, { Component } from 'react';
+import React, { useState, useRef, useEffect } from 'react';
 import { Provider } from 'react-redux';
-import CompassShell from './components/compass-shell';
 import CompassShellStore from './stores';
 import { getUserDataFilePath } from './modules/get-user-data-file-path';
 import { HistoryStorage } from './modules/history-storage';
 
 function createPlugin() {
   const store = new CompassShellStore();
-  const Plugin = class extends Component {
-    static displayName = 'CompassShellPlugin';
 
-    initializeHistoryStorage() {
-      const historyFilePath = getUserDataFilePath('shell-history.json');
+  function CompassShellPlugin() {
+    const [ShellComponent, setShellComponent] = useState(null);
+    const historyStorage = useRef(null);
 
-      if (!historyFilePath) {
-        return;
+    if (!historyStorage.current) {
+      const historyFilePath = getUserDataFilePath('shell-history.json');
+      if (historyFilePath) {
+        historyStorage.current = new HistoryStorage(historyFilePath);
       }
-
-      return new HistoryStorage(historyFilePath);
     }
 
-    /**
-     * Connect the Plugin to the store and render.
-     *
-     * @returns {React.Component} The rendered component.
-     */
-    render() {
+    useEffect(() => {
+      let mounted = true;
+
+      import(/* webpackPreload: true */ './components/compass-shell').then(
+        ({ default: Component }) => {
+          if (mounted) {
+            setShellComponent(Component);
```

Comment:
We already do, this splits one of the biggest parts of Compass into it's own chunk that Compass doesn't need to process during the critical startup time, less code needs parsing and evaluating on the startup, faster the application can start. This is the same reason I moved azure ip ranges into their own file, this was ~500Kb that would need to be processed before we can start the app.

I have plans to do the same for the `bson-transpiler` which is another big piece of Compass that we can safely move out of the application critical loading path.

In general we were discussing with Maurizio if its possible to do this to the whole plugin system, and even though it would make sense, this will be a much riskier change as currently hadron plugins are written with the expectation that everything is registered "synchronously" and before the app starts. Doing this on any level outside plugin core allow us to split app in separate chunks pretty safely and we don't really need much code upfront to start showing the connection screen.

Reply:
Gotcha, thanks for clearing it up, `webpackPreload ` makes sense now.